cdap-ui/app/cdap/services/datasource/index.js

```diff
@@ -0,0 +1,251 @@
+/*
+ * Copyright © 2016 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+import Socket from '../socket';
+import uuid from 'node-uuid';
+import Rx from 'rx';
+
+export default class Datasource {
+  constructor() {
+    let socketData = Socket.getObservable();
+    this.bindings = {};
+
+    this.socketSubscription =  socketData.subscribe(
+      (data) => {
+        let hash = data.resource.id;
+
+        if (!this.bindings[hash]) { return; }
+
+        if (data.statusCode > 299 || data.warning) {
+          this.bindings[hash].rx.onError(data.error || data.response);
+          if (this.bindings[hash].type === 'REQUEST') {
+            this.bindings[hash].rx.onCompleted();
+            this.bindings[hash].rx.dispose();
+            delete this.bindings[hash];
+          }
+        } else {
+          this.bindings[hash].rx.onNext(data.response);
+          if (this.bindings[hash].type === 'REQUEST') {
```

Comment:
I guess inner `if` condition is shared by both outer `if...else` condition structure.

We could separate this out something like this I guess,

```
if (data.statusCode > 299 || data.warning) {
  this.bindings[hash].rx.onError(data.error || data.response);
} else {
  this.bindings[hash].rx.onNext(data.response);
}
if (this.bindings[hash].type === 'REQUEST') {
  this.bindings[hash].rx.onCompleted();
  this.bindings[hash].rx.dispose();
  delete this.bindings[hash];
}
```